### ISSUE TITLE
issue with restart campaign

### ISSUE THREAD

**Issue by johbuch** [mautic/mautic#7338]
[//]: # ( Invisible comment: 
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Before you create the issue:
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Search for similar report among other reported issues.
Learn how to troubleshoot at https://www.mautic.org/docs/en/tips/troubleshooting.html
Use drag&drop to attach images or other files )

## Bug Description
If a contact that is in a segment that starts a campaign will be automatically added to the campaign. Then if this contact is removed from the segment, he will be automatically removed from the campaign.
Then if he is added back to the segment, he is not added back to the campaign even if the restart campaign option is activated.

| Q   | A
| --- | ---
| Mautic version | 
| PHP version | 
| Browser | 

### Steps to reproduce
1. create a dynamic segment with a filter
2. Create a contact and add the right value so that he can be in the segment
3. Create a campaign that starts on this segment and add activate option to allow contacts to restart the campaign
4. Run campaign.
5. Remove or change contact field value so that the contact is removed from the segment.
6. run the cron.
7. Add back the value and run the cron again and see that the contact is added back to the campaign
8. Check contact timeline and see that he has not been added a second time to the campaign
![image](https://user-images.githubusercontent.com/31535432/54672106-a65cfc00-4af7-11e9-8155-c6969cb0cdc1.png)

- Contact added to segment adresse
- contact added to campaign test
- contact remove from segment adresse
- contact remove from campaign test
- contact added to segment adresse
- this stops here

 
### Log errors

[//]: # ( Invisible comment:
Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable. )

**Comment by johbuch** [comment]
@alanhartless do you think it is a bug ?

**Comment by carloshmm** [comment]
My guess is that because the contact is removed prematurely from the campaign the "internals" of Mautic are still recognizing him as still belonging to that campaign. Maybe the contact needs to go all the way to the end of a campaign to trigger some internal mechanism that removes him from belonging to a campaign.
What I'm doing to solve this problem is:

Create a segment (segment A) that has the property conditions to enter the campaign.
Create another segment (segment B) with no properties.
Create campaign (campaign A) that has "segment A" as the entry point: In this campaign you just put an action to move the contact to "segment B".
Create campaign (campaign B) that has "segment B" as the entry point: This is the campaign that will do the actual work, at the end of this campaign add a action to remove the contact from "segment B".

Both "campaign A" and "campaign B" need to be recurring.

This is what I'm doing to solve this problem: https://github.com/mautic/mautic/issues/6439

It has worked so far.

**Comment by fankoti** [comment]
I have the same. The contact can not be returned to the campaign after removal from the segments. Contact added to segment, but not added to campaign. Campaign restart feature is activated. Mautic v2.15.1, PHP7.1

**Comment by mukomol** [comment]
@carloshmm can you explain littlebit more?

I try to use one campaign many times for one contact. For example when admin activate account or for user`s details notification about new order.

Example:
1. Create Segment A with condition (contact field 'block' == 0).
2. Create Segment B with no conditions.
3. Create Campaign A that only move contact to Segment B. Checkbox "Allow contacts to restart the campaign" set Yes.
4. Create Campaign B that send email "Your account unlocked" to contact . Checkbox "Allow contacts to restart the campaign" set Yes.

When some contact has been changed field 'block' set 0, it go through whole chain:
1. automaticaly added to Segment A
2. automaticaly added to Campaign A 
3. moved to Segment B
4. automaticaly added to Campaign B
5. Sent email
6. removed from Segment B

This works only one time. If I change contact fiald 'block' set 1, contact automaticaly removed from Segment A (failed the condition 'block' == 0) and removed from Campaign A.
Then if I change contact field 'block' to 0, contact added to Segment A, but does not added to Campaign A. Nothing happens more than one time.

Maybe I do something wrong? I know that we can move contact from campaign to campaign but how to catch event when contact field was changed in campaign, not in segment? I also know that we can reset campaign inside the campaign, make it loop, but it does not looks good.

**Comment by carloshmm** [comment]
> 1. Create Segment A with condition (contact field 'block' == 0).
> 2. Create Segment B with no conditions.
> 3. Create Campaign A that only move contact to Segment B. Checkbox "Allow contacts to restart the campaign" set Yes.
> 4. Create Campaign B that send email "Your account unlocked" to contact . Checkbox "Allow contacts to restart the campaign" set Yes.
> 
> When some contact has been changed field 'block' set 0, it go through whole chain:
> 
> 1. automaticaly added to Segment A
> 2. automaticaly added to Campaign A
> 3. moved to Segment B
> 4. automaticaly added to Campaign B
> 5. Sent email
> 6. removed from Segment B

When you say: "automatically", you are saying "after running the cron job mautic:segments:update" right?

The correct order is:
1. 'block' = 0 -> automaticaly added to Segment A (cron job mautic:segments:update inserts the contact on this segment)
2. automaticaly added to Campaign A (cron job mautic:campaigns:update inserts contact on this campaign)
3. moved to Segment B
4. End of Campaign A (There is no more steps on the campaign, the contact is removed from this campaign after executing mautic:campaigns:update)
5. automaticaly added to Campaign B (cron job mautic:campaigns:update inserts contact on this campaign)
6. Sent email
7. removed from Segment B
8. End of Campaign B (There is no more steps on the campaign, the contact is removed from this campaign after executing mautic:campaigns:update)

Continuing the process with restart of campaign:
9. 'block' = 1 (cron job mautic:segments:update removes contact from Segment A)
1. 'block' = 0 -> automaticaly added to Segment A (cron job mautic:segments:update inserts the contact on this segment)
... Continue from step 2 to 8 again.

Are you sure you are executing the cron jobs mautic:segments:update/mautic:campaigns:update/mautic:campaigns:trigger? You should execute them in order to make this to work.

**Comment by mukomol** [comment]
Hi, @carloshmm and thanks for answer.

First I checked cronjobs and logs then test mannualy sequentially launch few times
1. mautic:segments:update 
2. mautic:campaigns:update
3. mautic:campaigns:trigger

I get precisely I described above - scenario work out only one time. On second iteration when I change Block to 1 then back to 0 contact removed from Segment A, than added to Segment A, but does not added to Campaign A. There is no any errors in logs.

There is my segment A:
![image](https://user-images.githubusercontent.com/6931536/55136967-6372d800-5138-11e9-814e-bbbf74b9130e.png)

Segment B without any filters.

There is Campaign A:
![image](https://user-images.githubusercontent.com/6931536/55137274-3a9f1280-5139-11e9-9020-b8fe62cb4538.png)

And Campaign B:
![image](https://user-images.githubusercontent.com/6931536/55137296-4a1e5b80-5139-11e9-80af-d041f14becd2.png)

Mautic version is 2.15.1, changing contact field I do through REST API, if it makes matter.

I really can not understand where I make mistake. Have you any idea?

**Comment by carloshmm** [comment]
> I get precisely I described above - scenario work out only one time. On second iteration when I change Block to 1 then back to 0 contact removed from Segment A, than added to Segment A, but does not added to Campaign A. There is no any errors in logs.

Sorry for the delay. You are describing all the steps properly so I don't know why your campaign is not restarting. Are you sure your contact is leaving the campaigns? (Steps 4 and 8)

**Comment by Corsnec** [comment]
I have the same issue. With 2.15.0 we installed a process to controll our cronjobs. 

1. **Segment CronjobTest** inlcudes contacts with the **Tag Cronjob**
2. **Campaign A** starts with Contacts from **Segment CronjobTest**
3. **Campaign A** sends a Email to me
4. **Campaign A** puts Contact into Campaign B
5. **Campaign B** delets the **Tag Cronjob** then waits 12hours ( Now the Contact gets deleted automaticly from **Segment CronjobTest** & **from Campaign A**)  
6. **Campaign B** adds the  **Tag Cronjob** to the contact & aktive delets the Contact from **Campaign B**.

7. **Segment CronjobTest** adds the Contact
8. **Campaign A** should add the Contact but it doesnt. 

"Allow contacts to restart the campaign" is set to **yes**

To me it looks like a bug!

**Comment by pierrederuel** [comment]
Hi! Same problem here and that's a deal breaker for me! I cannot use Mautic without that kind of feature,  any update?

Thanks

**Comment by davidgut1982** [comment]
I wanna say that I can confirm this. I tried asking for help in [Slack,](https://mautic.slack.com/archives/C02HU8BUT/p1562678325494900) but came here to find it is not working in v2.16.0-dev.

**Comment by stevedrobinson** [comment]
Same issue here. totally hosed on a bunch of fronts due to this.

**Comment by npracht** [comment]
@johbuch can you test #7338 please ?

**Comment by johbuch** [comment]
tested PR #7704 --> seems to fix this issue 👍

**Comment by pety-dc** [comment]
When a contact is removed from a segment it gets removed from all campaigns that use that segment as source.

I think the reason why campaign is not restarted on a contact if the contact gets readded to the segment is that the contact previously was removed from the campaign with "manually=yes" flag.

Contacts that were manually removed from a campaign don't get readded.


`Mautic\CampaignBundle\Membership\Action\Remover`

```
public function updateExistingMembership(CampaignMember $campaignMember, $isExit)
    {
        if ($isExit) {
            // Contact was removed by the change campaign action or a segment
            $campaignMember->setDateLastExited(new \DateTime());
        } else {
            $campaignMember->setDateLastExited(null);
        }

        if ($campaignMember->wasManuallyRemoved()) {
            $this->saveCampaignMember($campaignMember);

            // Contact was already removed from this campaign
            throw new ContactAlreadyRemovedFromCampaignException();
        }

        // Unschedule any scheduled events
        $this->leadEventLogRepository->unscheduleEvents($campaignMember, $this->unscheduledMessage);

        // Remove this contact from the campaign
        **$campaignMember->setManuallyRemoved(true);**
        $campaignMember->setManuallyAdded(false);

        $this->saveCampaignMember($campaignMember);
    }
```

Why is it necessary to apply segment removal as _manual_ ?

**Comment by StevenChin** [comment]
@johbuch I was having the same issue as well. I was also able to reproduce the issue using the change campaign action in the campaign builder to remove contacts from the campaign.

I did some investigation into how the `mautic:campaign:rebuild` command works and found that it was designed to only allow **new contacts** to enter campaigns - or you can manually add contacts to the campaign. Contacts will only be added to a campaign when they are in the campaign's segment but are NOT in the `campaign_leads` table, see [updateQueryWithExistingMembershipExclusion()](https://github.com/mautic/mautic/blob/2.15.3/app/bundles/CampaignBundle/Entity/LeadRepository.php#L604) for more details.

#7704 works in both cases (when contacts are removed from campaigns via segment change or campaign action) because it updates the aforementioned check to find contacts that were manually removed from Campaigns that can be restarted or contacts that don't exist in the `campaign_leads` table.

**Comment by ttencate** [comment]
The above patch (https://github.com/medullan/mautic/pull/1/files) does not work for me, and I think it's more or less because of what @pety-dc describes.

Indeed, I see `manually_removed = 1` in the `campaign_leads` table. However, note that the code snippet quoted by @pety-dc is about membership of _campaigns_, not _segments_. In the `lead_lists_leads` table, I see `manually_removed = 0` as expected.

So the more technically correct question is: _Why is it necessary to apply removal from a **campaign** as _manual_?_

**Comment by dennisameling** [comment]
PR #7704 is now merged for the 2.16 release, so I'm closing this issue. :)
